### PR TITLE
First pass at basic query string parameters.

### PR DESCRIPTION
**Pull request by mitchdenny**
# First pass at basic string binding for request delegate source generator

This PR gets the basics of binding strings to the source generated request delegate.  It builds on the special type parameter handling. The PR includes logic to detect the IFromX metadata interfaces and set the parameter type accordingly (this is the explicit handling logic discussed.

It includes partial null handling in that it will return a 400 if StringValues is null or empty - but it doesn't do full on nullability handling around things like nullable arrays, arrays of nullables, and nullable arrays of nullables.

You can see where I've got a pattern matching switch to detect what kind of parameter preparation logic to emit. The idea here is that if we have a particular special case which is messy to handle with just one block of code we can use this switch to provide alternative implementations, but mostly it'll just switch on the Source property.

(note: some tests aren't passing yet because I haven't updated the baselines - that'll be the last thing I do)

### CODE REVIEW

**Comment by mitchdenny** [comment]
@captainsafia here is the minimal PR that I think will help us avoid stepping on each others toes.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:39]:
```suggestion
        if (parameter.Type is INamedTypeSymbol parameterType && parameterType.NullableAnnotation == NullableAnnotation.Annotated)
```

Also, I would stick the logic for setting `IsOptional` inside the case for the parameter itself. There's enough subtleties between optionality for different parameter types:

- Body parameters needs to handle the `AllowEmpty` property
- `BindAsync` parameters need to handle distinctions in nullability annotation between the parameter and the value returned from `BindAsync`

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:126]:
```suggestion
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_NoParam_StringReturn_WithFilter.generated.txt
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_NoParam_StringReturn_WithFilter.generated.txt:113]: Nit: indentation.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_WithParams_StringReturn.generated.txt
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_WithParams_StringReturn.generated.txt:13]: 🤩

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:40]: Add a test for the nullability scenario since the code seems to handle that?

File: src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj
Comment (davidfowl) [inline src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj:19]: ???

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:7]: These new usings look suspect

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt:117]: && true? What is that for?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt
Comment (davidfowl) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt:133]: Extra whitespace 😢

File: src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj
Comment (mitchdenny) [reply src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj:19]: We need the ```[FromQuery]``` attributes. The test cases basically copy the dependencies of the unit tests project that they are defined within.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs
Comment (mitchdenny) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:7]: Not even sure how they got in there! Removing.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt
Comment (mitchdenny) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt:117]: It makes a little more sense when you look at the template that generates. Basically on a per endpoint parameter basis we sometimes want to opt into return a 400 status code when ```IsOptional``` is true - but that is evaluated at the time source gen time not at runtime. I could probably change this to exclude the whole block rather than doing this.

I'll fix it up.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt
Comment (mitchdenny) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt:133]: I'll see if I can figure out where that extra one is coming from. That said I think we'll drive ourselves nuts trying to track down errant new lines as we iterate on this thing. I think it would be better to just do a blanket code-format before we spit the code out at the very end rather than trying to perfect line feeds and indentation.

Comment (mitchdenny) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_MultipleStringParam_StringReturn.generated.txt:133]: To give you a concrete example of what sweating a bit of whitespace is going to do:

```csharp
        for (var parameterIndex = 0; parameterIndex < endpoint.Parameters.Length; parameterIndex++)
        {
            var parameter = endpoint.Parameters[parameterIndex];

            var parameterPreparationCode = parameter switch
            {
                {
                    Source: EndpointParameterSource.SpecialType
                } => parameter.EmitSpecialParameterPreparation(),
                {
                    Source: EndpointParameterSource.Query,
                } => parameter.EmitQueryParameterPreparation(),
                _ => throw new Exception("Unreachable!")
            };

            // To avoid having two newlines after the block of parameter handling code.
            if (parameterIndex < endpoint.Parameters.Length - 1)
            {
                parameterPreparationBuilder.AppendLine(parameterPreparationCode);
            }
            else
            {
                parameterPreparationBuilder.Append(parameterPreparationCode);
            }
        }
```

The last if block is just noise from a maintainability point of view.

**Review by wtgodbe — CHANGES_REQUESTED** [review]
Please undo the submodule changes

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:113]: I assume this is intentionally since you're anticipating that I'll push this forward with my changes to introduce the variadic `InvocationFilterContext` overloads.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt:231]: Nit: I assume this empty whitespace is from the empty parameter preparation block since this handler takes no parameters.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_SingleNullableStringParam_WithoutQueryStringValueProvided_StringReturn.generated.txt
Comment (captainsafia) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_SingleNullableStringParam_WithoutQueryStringValueProvided_StringReturn.generated.txt:115]: Nice use of the comments here!

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs
Comment (captainsafia) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:45]: It looks like the pattern here is going to be slightly different than what we currently do in RDF. In RDF, we assign a `wasParamCheckFailure` variable to track if any binding failed then set the status code and return _after_ all the bindings have been attempted.

I'm fine with merging this as-is now, since I:

- Anticipate (hope?) that people aren't relying on side-effects from parameter-binding in their implementations
- I think we'll end up bringing back a `wasParamCheckFailure` variable as we add support for more binding here to keep the generated code more terse

**Review by captainsafia — COMMENTED** [review]
LGTM for the generator changes

It looks like there have been some unintended changes made to the git submodules. I'd revert those before merging.

**Review by captainsafia — APPROVED** [review]
Approving since I'm fine with the generator changes and @wtgodbe has already requested changes for the submodule changes I was nervous about. 😅

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt
Comment (mitchdenny) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt:231]: Yep. I think it would make the emission code less readable if we tried to remove this :)

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs
Comment (mitchdenny) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:113]: Yep, for now we didn't need the argument list, but I left it as an extension method even though technically it doesn't use any of the state from the target object.

**Comment by mitchdenny** [comment]
> Please undo the submodule changes

Done. Sorry about that. GitHub was giving me issues yesterday not recognizing a push that I had made, and I did a sloppy whitespace change and ended up pulling in the submodules update inadvertently.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs
Comment (mitchdenny) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointParameterEmitter.cs:45]: Yeah, I think once this is in, we can come up with a pattern for this. It is going to be "fun" replicating existing behaviors and quirks.

**Comment by dougbu** [comment]
> > Please undo the submodule changes
> 
> Done. Sorry about that. GitHub was giving me issues yesterday not recognizing a push that I had made, and I did a sloppy whitespace change and ended up pulling in the submodules update inadvertently.

`git config --global --add submodule.recurse true` works wonders and will affect all future aspnetcore clones you make.

File: src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj
Comment (wtgodbe) [reply src/Http/Http.Extensions/test/Microsoft.AspNetCore.Http.Extensions.Tests.csproj:19]: We'll sometimes copy a test helper class between projects, or move it to `src/Shared`, but I don't think it's worth blocking the PR on that. Maybe as a follow-up.

**Review by wtgodbe — APPROVED** [review]

**Pull request merged**
